### ISSUE TITLE
Heydon: various

### ISSUE THREAD

**Issue by Heydon** [adamsilver/f#110]
> Apart from the numbers, there are many reasons why it drastically improves the user experience.

The numbers aren't a reason; they're evidence of a reason. How about *"Behind the improved numbers, there are many reasons why it drastically improves the user experience."*

> That is not to say you’ll always end up combining screens — in my experience the best results come from splitting things apart, period.

I don't think you need to say this. It's implicit since you already started with "one thing per page" then qualified that _sometimes_ grouping is better.

> Conversely, a long form takes a long time to complete.

Hmmm. A form divided up over pages still takes long, but it seems less long? I think you already covered the part about incremental submission, so not sure this bullet is needed except to say the page being open a long time has the freeze/crash risk.

> It allows you to innovate as the entire screen is dedicated to one thing.

Not sure what you mean by this. Innovate what? Affords you more space in which to provide supplementary content and create aesthetic nuances?

![email field](https://user-images.githubusercontent.com/683558/33382435-a69ca78e-d518-11e7-9c74-849655c04e05.png)

> (Note: the button's positioning and styling is the same as set out in “A Registration Form”. But the label is set to “Continue” which implies progress and is better suited to the linear checkout flow).

I found this confusing because it's referring to something that isn't in the image despite appearing under it, like a caption.

![mobile field](https://user-images.githubusercontent.com/683558/33382507-d7bd5764-d518-11e7-961b-4798c3a6be60.png)

I'm wondering whether the omission of the terms "phone" and "number" here is a mistake. It feels conspicuously missing and some users with cognitive issues may be thrown?

![mobile screenshot with telephone label](https://user-images.githubusercontent.com/683558/33382699-562a541c-d519-11e7-8f71-d3768dced7cd.png)

Then, in the above image, the label changes to "telephone" (tel). I appreciate that you're using a test screen for input types here, but it can be misunderstood as an extrapolated version of the previously discussed inputs and just seem inconsistent.

> So we follow principle 5, Offer choice, by marking the field as optional.

An I noted in **Book A Flight**, this kind of thing is a misapprehension of the **Offer choice** principle. **Offer choice** is about choices over how to complete tasks, not the ability to make choices per se. Maybe the principle is unclear!

> including the phrase ‘optional’ after a label is much clearer than any visual symbol you could use to mean the same thing.

+1 for '(optional)' over required asterisks! I do think `aria-required="true"` on required fields makes things clearer and could reduce errors among screen reader users, though.

Example:

```html
<div class="field">
  <label for="address1">
  	<span class="field-label">Address line 1</span>
  </label>
  <input type="text" id="address1" name="address1" aria-required="true">
</div>
```

> If no address is found, users can change the interface back to the original address form. In doing so, we conform to principle 5, offer choice.

That's what **Offer choice** is about! (In this case: assisted versus manual entry)

> In most screen readers, the legend is read out in combination with the radio button's label which in this case would be “Delivery options, Standard (Free, 2-3 days)”.

Notably, in NVDA and JAWS (both, I think, but might need testing) the group label is only read out along with the first individual field that is focused when entering the form (in either direction). Just tested and Safari + Voiceover supports `<legend>` but announces it for each field. In Chrome + Voiceover there is no support. Fortunately, this is an unusual combination of browser/screenreader.

> The container has a role attribute set to status and aria-live set to polite. When the status box is updated, screen readers will announce it, but only after the user finishes typing. This way, users aren't rudely interrupted.

Reminder that we need to test this. I _think_ some screen reader / browser combos _will_ interrupt, so there may be issues that might need addressing with debouncing or just activating the live region when the number of remaining characters becomes critical.

> Design is a team sport

Citation?

> we use them to in dates

*we use them in dates

> In these cases, it's safer to use a regular text box to avoid excluding users unnecessarily. Remember users are still able to type numbers this way it's just that the buttons are smaller. To further soften the blow, the numeric keyboard can be triggered for iOS users by using the pattern attribute.

Just to say the stuff about `type="number"` and inconsistencies/inappropriateness is very interesting and helpful 👏

> And to top it off, the number is never beside a label to demarcate it making the field rather ambiguous.

Not sure what you mean here. Maybe there should be a comma after 'demarcate it'? In any case, is there really _never_ a label? Isn't CVC/CVV the label? Do you mean not often / never a qualifying label? Your suggestion of a hint is good, but this lead-up is just a bit difficult to follow.

> To do this, we need to listen to the checkbox's click event.

Isn't it better to listen to the `change` or `input` events?

```js
function CheckboxCollapser(checkbox, container) {
  this.checkbox = checkbox;
  this.container = $(container);
  $(this.checkbox).on('click', $.proxy(this, 'onCheckboxClick'));
  this.check();
};
```

I know I've mentioned this before, but the `this`/$proxy/prototype wrangling camouflages the intent of the script for me. It took me a long time to realize that `$.proxy(this, 'onCheckboxClick')` was kind of equivalent to a closure. Overall, for me a simple script divided over one function and two prototype methods feels overkill. But I might be missing something. 

Given the checkbox as `checkboxElem` and the container as `containerElem`, I'd probably just do something like 

```js
function toggleAddress (checkbox, container) {
  container[checkbox.checked ? 'hide' : 'show']();
}

checkboxElem.addEventListener('click', function() {
   toggleAddress(checkboxElem, containerElem)
})
```

> adding an extra step in the flow reduces friction

Mentioned before, but I think 'friction' is rather jargony. What is meant by friction?

> Take Mary (I made her up), a mother of two infants. It's the middle of the night, and her baby is crying inconsolably. Naturally, Mary is tired and stressed. To make things worse, she's run out of nappies.

I'd much prefer we had the parent as the father here. Just because it's always the assumption that the woman takes care of the babies, and Mary feels like a bit of a stereotype.

> Even the most sophisticated validation pattern cannot eradicate human error.

Not _all_ human error. Need to be specific about what is valid and what is just 'not what I meant'.

> This is likely to result canceled orders.

*Result _in_ 

![review screen](https://user-images.githubusercontent.com/683558/33389217-0a0c90d4-d52a-11e7-9c02-7b3b63faadfe.png)

Probably a good idea to mention each editable line should belong to a list as a `<li>`.

*TODO: nearly there!*

> Up to now, we've also been sure to use plain and simple language for labels, hints, and errors. On the confirmation page, there is an opportunity to let your brand's personality shine through.

Perhaps spell out why it's okay to be more whimsical here (the impirtant stuff has already been done).

> All in all, this reduces friction to nil and significantly improves conversion. Remember Mary from earlier: the next time she runs out of nappies, buying them should be a breeze, something I'm sure she'd appreciate if she was a real human being.

> In which case, we can bypass almost every step of the checkout flow including, email, mobile, delivery address and options and of course payment. Instead, we can take users straight to the Review page. This way users get a reminder of their default preferences with the chance to make amends.

I feel like the amendment interface could do with an illustration.

Fucking hell 😂 Yeah — I think this persona needs some work not to (accidentally) offend anyone.
There's a great opportunity to make a nappy buyer a man here, I'd take it! Also, "if she/he was a real human being" is redundant and could be catastrophically misconstrued...

> Progress bars are often used within checkout because they give users an idea of where they are in the process and how long's left. While this makes sense, there's little data that shows including one adds value.

I'd like to see the progress bar illustrated somehow. Also, that last sentence doesn't scan for me?

> To top it off, handling conditional sections is hard because you either always show the step which is misleading, or you show it when it's relevant which is dishonest.

Again, not really grasping this. An illustration would help. Not sure if 'dishonest' is the right word here. Makes it sound unethical, but I think it's just confusing?

> Instead, it's better to start without a progress bar. Then you can test to see if your users struggle without one by conducting research. Remember, it's far easier (and cheaper) to add features than it is to remove them later on.

I don't think you go the progress bar enough. What would it look like? How would you mark it up to be accessible? Off the top of my head, it would probably have to incorporate some text like "65% complete" — possibly hidden visibly? 

This section feels like a bit of a stub overall, and I think illustrations should abound — you're talking about the layout after all.

> Giving users this ability digitally is of similar important.

Needs rephrasing ("importance"?)

> Whatever you do, don't place it before the submit button. When keyboard users press Tab when inside the last field, pressing Enter will take them back to the previous step instead of submitting the form which is frustrating.

Not sure this is a problem since they can see or hear that the newly focused control is "back", not "submit" — but I agree about placing it top left.

**Done!** Lots of great stuff here, but I think some work is needed to fill it out towards the end.
